pkg/filter/validator/basicauth.go

```diff
@@ -0,0 +1,168 @@
+// Only one of UserFile or UseEtcd should be defined.
+type BasicAuthValidatorSpec struct {
+	// UserFile is path to file containing encrypted user credentials in apache2-utils/htpasswd format.
+	// To add user `userY`, use `sudo htpasswd /etc/apache2/.htpasswd userY`
+	// Reference: https://manpages.debian.org/testing/apache2-utils/htpasswd.1.en.html#EXAMPLES
+	UserFile string `yaml:"userFile" jsonschema:"omitempty"`
+	// When UseEtcd is true, verify user credentials from etcd. Etcd stores them:
+	// key: /credentials/{user id}
+	// value: {encrypted password}
+	UseEtcd bool `yaml:"useEtcd" jsonschema:"omitempty"`
+}
+
+// BasicAuthValidator defines the Basic Auth validator
+type BasicAuthValidator struct {
+	spec                 *BasicAuthValidatorSpec
+	authorizedUsersCache *lru.Cache
+	cluster              cluster.Cluster
+}
+
+func parseCredentials(creds string) (string, string, error) {
+	parts := strings.Split(creds, ":")
+	if len(parts) < 2 {
+		return "", "", fmt.Errorf("bad format")
+	}
+	return parts[0], parts[1], nil
+}
+
+func (bav *BasicAuthValidator) getFromFile(targetUserID string) (string, error) {
+	file, err := os.OpenFile(bav.spec.UserFile, os.O_RDONLY, os.ModePerm)
+	if err != nil {
+		return "", fmt.Errorf("open file error: %v", err)
+	}
+	defer file.Close()
+	sc := bufio.NewScanner(file)
+	for sc.Scan() {
+		line := sc.Text()
+		userID, password, err := parseCredentials(line)
+		if err != nil {
+			return "", err
+		}
+		if userID == targetUserID {
+			return password, nil
+		}
+	}
+	return "", fmt.Errorf("unauthorized")
+}
+
+func (bav *BasicAuthValidator) getFromEtcd(targetUserID string) (string, error) {
+	const prefix = "/credentials/"
+	password, err := bav.cluster.Get(prefix + targetUserID)
+	if err != nil {
+		return "", err
+	}
+	if password == nil {
+		return "", fmt.Errorf("unauthorized")
+	}
+	return *password, nil
+}
+
+// NewBasicAuthValidator creates a new Basic Auth validator
+func NewBasicAuthValidator(spec *BasicAuthValidatorSpec, supervisor *supervisor.Supervisor) *BasicAuthValidator {
+	var cluster cluster.Cluster
+	if spec.UseEtcd {
+		if supervisor == nil || supervisor.Cluster() == nil {
+			logger.Errorf("BasicAuth validator : failed to read data from etcd")
+		} else {
+			cluster = supervisor.Cluster()
+		}
+	}
+	cache, err := lru.New(256)
+	if err != nil {
+		panic(err)
+	}
+	bav := &BasicAuthValidator{
+		spec:                 spec,
+		authorizedUsersCache: cache,
+		cluster:              cluster,
+	}
+	return bav
+}
+
+func (bav *BasicAuthValidator) getUserFromCache(userID string) (string, bool) {
+	if val, ok := bav.authorizedUsersCache.Get(userID); ok {
+		return val.(string), true
+	}
+	var password string
+	var err error
+	if bav.spec.UserFile != "" {
+		password, err = bav.getFromFile(userID)
+	}
+	if bav.spec.UseEtcd == true {
+		password, err = bav.getFromEtcd(userID)
+	}
+	if err != nil {
+		return "", false
+	}
+	bav.authorizedUsersCache.Add(userID, password)
+	return password, true
+}
```

Comment:
The data in Etcd may change, in this case, we cannot use the cached data.

Reply:
We could listen for changes in etcd (or htpasswd file) and invalidate cache keys that are not longer present. What do you think?

Reply:
yes, we need to watch the changes, there's a `Syncer` in the cluster package, which can be used for this.